### PR TITLE
check-dispatcher refactor and es6 conversion

### PR DESCRIPTION
**Pull request by GHaberis**
## What does this change?

- Converts check-dispatcher to es6.

- Moves 'dependent check' logic out of check-mediator-checks and into check-dispatcher. Moving this logic into single place (check-dispatcher) makes reasoning about what checks are doing easier, rather than splitting logic across multiple files.

- Updates outbrain.js to run 3 checks isUserInContributionsAbTest, isUserInEmailAbTestAndEmailCanRun and isStoryQuestionsOnPage individually rather than running single test isOutbrainNonCompliant. Running these checks individually means we can track the actual cause of a non-compliant outbrain widget more granularly.

- Update outbrain-load.js to track state of outbrain component.

 What is the value of this and can you measure success?

- More es6
- Consolidates check-mediator/dispatcher logic into single location
- Simplifies check-mediator as no longer needs to manage dependent checks
- Allows us to track outbrain state more granularly

## Does this affect other platforms - Amp, Apps, etc?

No

## Screenshots

N/A

## Tested in CODE?

No

### CODE REVIEW

**Comment by GHaberis** [comment]
@guardian/commercial-dev I've made some updates to the check-mediator/dispatcher that determine the state of outbrain in this PR. The decision making logic remains the same, where it sits has moved around a little, it's hopefully easier to follow now. The changes made here will allow us to track what condition/check resulted in a non-compliant outbrain widget at a more granular level.

File: static/src/javascripts/projects/commercial/modules/third-party-tags/outbrain-tracking.js
Comment (SiAdcock) [inline static/src/javascripts/projects/commercial/modules/third-party-tags/outbrain-tracking.js]: Since this field is already optional, I don't think it makes sense for it to be nullable as well. Instead of explicitly setting this to `null`, why don't we simply omit the property altogether? Does tracker-js (or Ophan itself) behave differently depending on whether this property is `null` or `undefined`?

File: static/src/javascripts/projects/commercial/modules/third-party-tags/outbrain.spec.js
Comment (SiAdcock) [inline static/src/javascripts/projects/commercial/modules/third-party-tags/outbrain.spec.js]: Seems redundant, given the next line 😄

File: static/src/javascripts/projects/common/modules/check-dispatcher.js
Comment (SiAdcock) [inline static/src/javascripts/projects/common/modules/check-dispatcher.js]: Are we going to add Flow type annotations as part of this conversion?

File: static/src/javascripts/projects/common/modules/check-dispatcher.js
Comment (gustavpursche) [inline static/src/javascripts/projects/common/modules/check-dispatcher.js]: `return Promise.all(dependentChecks).then(everyCheckPassed);`. For the following occurences too.

File: static/src/javascripts/projects/commercial/modules/third-party-tags/outbrain-tracking.js
Comment (GHaberis) [inline static/src/javascripts/projects/commercial/modules/third-party-tags/outbrain-tracking.js]: @SiAdcock yes Ophan itself behaves differently if widgetId isn't present, it ignores requests without widgetId right now. 

In order to track the info we need I'm going to need to raise a PR against ophan/ophan-data-lake to handle widgetId as null but yes it's probably cleaner to make it optional. 

I'll get the PR in order for this repos before merging this.

**Review by gu-stav — DISMISSED** [review]
Minor thing plus what Simon said. Good job 🌷

File: static/src/javascripts/projects/common/modules/check-dispatcher.js
Comment (GHaberis) [inline static/src/javascripts/projects/common/modules/check-dispatcher.js]: I wasn't sure what the suggested practise was regarding types, I followed the approach that they're mandatory on public functions then optional on private functions? I'm happy to add type annotations if you'll feel they'll provide useful guidance for developers editing/reading this file!

**Comment by regiskuckaertz** [comment]
@SiAdcock @gustavpursche Are we allowed to use `async`/`await`? Part of the reason why this module is hard to read is the promise chain, and I wonder if it would make our lives easier if it was written in imperative style. Not sure I got this right, but my attempt [here](https://gist.github.com/regiskuckaertz/3c34632d36874614e485efc8a74c5e6a).

File: static/src/javascripts/projects/commercial/modules/third-party-tags/outbrain.spec.js
Comment (GHaberis) [inline static/src/javascripts/projects/commercial/modules/third-party-tags/outbrain.spec.js]: Can't really argue with this!

**Comment by SiAdcock** [comment]
@regiskuckaertz 

From what I understand, in order to transpile async functions, we would need to use Babel's [`transform-async-to-generator`](https://github.com/babel/babel/tree/master/packages/babel-plugin-transform-async-to-generator) plugin. In turn, this requires us to serve a [generator runtime](https://github.com/facebook/regenerator/blob/master/packages/regenerator-runtime/runtime.js) (or [Babel's polyfill library](https://babeljs.io/docs/usage/polyfill/)), which adds a lot of weight to the application.

We could potentially spike out how much extra weight this would add. Or @sndrs may be able to shed further light on this subject.

File: static/src/javascripts/projects/common/modules/check-dispatcher.js
Comment (SiAdcock) [inline static/src/javascripts/projects/common/modules/check-dispatcher.js]: As we discussed IRL, I don't think there's any distinction in type annotation requirements between public and private functions. It's a case of balancing time you want to spend versus the [piece of mind that Flow gives us](https://github.com/guardian/frontend/wiki/So-you-want-to-ES6%3F#specific-way) 😅

**Comment by GHaberis** [comment]
@SiAdcock I've removed the references to 'state' in the tracking calls as the Ophan work to support this new property needs to be done before this data can be recorded. 

Rather than blocking this PR I'll raise a new PR to add 'state' to the tracking calls once Ophan is ready to accept it.

File: static/src/javascripts/projects/commercial/modules/third-party-tags/outbrain.js
Comment (SiAdcock) [inline static/src/javascripts/projects/commercial/modules/third-party-tags/outbrain.js]: How do you feel about adding return / param types to these functions?

File: static/src/javascripts/projects/common/modules/check-dispatcher.js
Comment (SiAdcock) [inline static/src/javascripts/projects/common/modules/check-dispatcher.js]: Just noticed the signature of this and `hasLowPriorityAdLoaded` are inconsistent with other asynchronous checks in here. It might be dangerous to sometimes return a promise and sometimes not, as the calling function can't rely on the return value being `thennable`. Is it possible to make these function always return a `Promise<boolean>`, for safety?

Comment (SiAdcock) [inline static/src/javascripts/projects/common/modules/check-dispatcher.js]: To go further, perhaps all checks should always return `Promise<boolean>` even if they are not async, for consistency?

File: static/src/javascripts/projects/common/modules/check-mediator.js
Comment (SiAdcock) [inline static/src/javascripts/projects/common/modules/check-mediator.js]: If this function initially set `registeredChecks = {}`, we wouldn't need to expose the `testClean` function 🎉

Comment (gustavpursche) [inline static/src/javascripts/projects/common/modules/check-mediator.js]: Nice idea!

File: static/src/javascripts/projects/commercial/modules/third-party-tags/outbrain.js
Comment (GHaberis) [inline static/src/javascripts/projects/commercial/modules/third-party-tags/outbrain.js]: Yep i'm happy to do this 👍

File: static/src/javascripts/projects/common/modules/check-dispatcher.js
Comment (GHaberis) [inline static/src/javascripts/projects/common/modules/check-dispatcher.js]: The return values of these checks don't have to be `thennable`, their return value returned is passed as an argument to the `resolve` function for the promises defined in check-mediator. These promises can be resolved with a boolean or another promise.

**Comment by GHaberis** [comment]
@SiAdcock I've addressed these outstanding issues now.

- added types to check-dispatcher
- return type of all checks in dispatcher is now a Promise
- remove _ test function in check-mediator.

**Review by SiAdcock — APPROVED** [review]
Love it!

**Pull request merged**
